Syntax sugar: user-defined literals for arg and str

dean0x7d opened the pull request:
Pure sugar here. This PR proposes using C++11 [UDLs](http://en.cppreference.com/w/cpp/language/user_literal) for `py::arg` and `py::str`. Here is an example for named arguments:
```c++
// regular version
m.def("foo", &foo, py::arg("required"), py::arg("optional")=value);
// using literals
m.def("foo", &foo, "required"_a, "optional"_a=value);
```
where `_a` is the proposed literal suffix for `py::arg`. I intentionally omit the spaces in `_a=value` as this matches the usual Python style for default values.

This is similar for `py::str` with the `_s` suffix:
```c++
py::set set;
set.add(py::str("key1"));
set.add("key2"_s);
```

Some notes:

* The literals are in an inline namespace as is the `std` library convention. Before use, they must be imported with `using namespace pybind11::literals`. This will bring in only the `_a` and `_s` names, but nothing else from the `pybind11` namespace. Alternatively, `using namespace pybind11` can be used which will import everything including the literals thanks to the inline namespace.

* I wasn't sure if I should mix the inline keyword with the existing `NAMESPACE_BEGIN/END` macros, so I just wrote the plain `namespace {}`. I'm also not quite sure why the namespace macros are needed. Let me know if I should change anything here.

* I haven't added anything to the docs yet (in case the feature isn't desired) but I can add a little note for the literals.

wjakob commented:
I love it! Yes, please add something to the docs (perhaps the first time ``arg`` is used in the basics part of the documentation? I would prefer it if you leave the first ``py::arg`` occurrence but then point out that there is also a shorter notation using user defined literals, as opposed to killing py::arg everywhere.)

With regards to why ``NAMESPACE_BEGIN`` / ``END`` is used instead of a plain ``namespace x { };``, it's just to avoid some braces at the top level that can throw off editors which want to indent the whole file by 4 spaces.

wjakob commented:
This (https://github.com/dean0x7d/overload_cast) is really neat by the way. Do you see any way to have a uniform syntax between the C++11 and C++14 versions?

dean0x7d commented:
> This (https://github.com/dean0x7d/overload_cast) is really neat by the way. Do you see any way to have a uniform syntax between the C++11 and C++14 versions?

I was hoping to, but I think the `overload<int>{}(foo)` version is the best that can be done in C++11 (macros don't save it either). I had an alternative idea which would need to be partially integrated with `def`, but I'm not sure how intrusive it would be. I'll get back to you with a concrete proposal.

> Yes, please add something to the docs (perhaps the first time arg is used in the basics part of the documentation? I would prefer it if you leave the first py::arg occurrence but then point out that there is also a shorter notation using user defined literals, as opposed to killing py::arg everywhere.)

Yeah, I was also thinking right around [here](http://pybind11.readthedocs.io/en/latest/basics.html#keyword-arguments) would be a good place to mention it, but only as the alternative syntax. But I'm not sure where to mention `str` and `_s`. Any suggestions?

wjakob commented:
Sounds good, keep me posted :).

The wrapper classes like ``str``, ``int_`` are barely documented at this point (the only mention is advanced.rst around line 1266).

Thinking a bit about this, I wonder how important ``str_`` UDL is (vs arg). It would be a bit odd if there is a literal for ``str_``, but not ``int_``, ``float_``, etc.

How about we leave this one out for now and just focus on ``arg_``, which will occur much more frequently.

dean0x7d commented:
I agree about `str`. I removed that commit.

I added the `_a` docs. It may be that I'm over-explaining a simple concept there, but it seems like UDLs are still are relatively obscure feature in C++.

wjakob commented:
This is so cool! And all agreed, it is an obscure feature -- I never had a clue that it could do this sort of stuff.
